File: n_rule_maps.py
import math
import matplotlib.pyplot as plt
import random
import numpy as np

def lineAngle(slope1, slope2):
    #for angle between two lines use
    ang = math.atan(abs((slope2-slope1) / (1 + (slope2*slope1))))
    return ang

def lineZero(m, b):
    # find the root of the passed line, want when y=0
    x = (-b)/m
    y = 0
    return x, y

def findFunc(point1, point2):
    """find the linear function that fits two points"""
    m = ((point2[1] - point1[1]) / (point2[0] - point1[0]))
    b = (((point2[0] * point1[1]) - (point1[0] * point2[1])) / (point2[0] - point1 [0]))
    
    return m, b


def pFunc(m):
    """find function perpindicular to passed function slope"""
    if(m != 0):
        mperp = (-1/m)
        
    if(m == 0):
        mperp = 0
        
    return mperp

def newB(m, point):
    """find the y-intersection that the function/point pair passes through"""
    b = (point[1] - (m * point[0]))
    return b

def dist(point1, point2):
    """find the distance between the two passed points"""
    dist = math.sqrt((pow((point2[0] - point1[0]), 2)) + ((pow((point2[1] - point1[1]), 2))))
    return dist

def inter(mb1, mb2):
    """find the point of intersection between two lines"""
    x = ((mb2[1] - mb1[1]) / (mb1[0] - mb2[0]))
    y = ((mb1[0] * x) + mb1[1])
    return x, y

def dist_point_line(line, point):
    """return the distance between a point of form [x, y] and line of form [slope, intercept]"""
    # first find orthogonal line
    s = math.tan(math.radians(line[0]))
    p = pFunc(s)
    perp_line = [p, newB(p, point)]
    # now find intersection of to lines, where lp is converted from angle to usual slope
    lp = (s, line[1])
    # find line intersection point
    intersection = inter(perp_line, lp)
    # determine distance and return
    d = dist(point, intersection)
    return d

def mapping_options(target_line_angle, step_angle):
    """takes the target line slope angle, and the given rule step angle, and returns the two lines
    that intersect at the target line at the step angle"""
    
    # first get base step angle slopes
    base = [math.tan(math.radians(step_angle)), math.tan(math.radians(-step_angle))]
    
    # now construct other lines and give data, in form [line angle 1, line angle 2]
    lines = [math.tan(math.radians(step_angle + target_line_angle)), math.tan(math.radians(target_line_angle - step_angle))]
    return lines

def plot(dpi):
    fig, ax=plt.subplots(dpi=dpi)
    ax.legend()


# find indices of list duplicates
def list_duplicates_of(seq,item):
    start_at = -1
    locs = []
    while True:
        try:
            loc = seq.index(item,start_at+1)
        except ValueError:
            break
        else:
            locs.append(loc)
            start_at = loc
    return locs

def nRuleMap(X_m, nrule, num_iter, points_to_plot, start_point, orbit_color="red"):
    """The following code takes an n-rule, and plots the path it generates
    in the specified space X_m.  If the n-rule map converges to a periodic orbit
    within num_iter iterations, the program will return the period.
    Inputs:
    X_m: input a list of m tuples, where m is the number of lines plus the x axis, so 
        X_m is of form [(slope1, intercept1), (slope2, intercept2), ...] for line slope and intercept
        where line slope is an angle in degrees.  Further, we require that len(X_m) >= 2
    n-rule: takes a list of n length three tuples of form 
        [(angle1, orientation1, distance1), (angle2, orientation2, distance2), ...]
        where each angle parameter is between 0 and 90, orientation is either 1 or 0, and distance 
        is a positive integer from 1 to m inclusive.  If angle=90, orientation can be either 0 or 1 with no change
        of map behavior
    num_iter: number of times to iterate the map
    points_to_plot: an interval [start_point, end_point] that determines the interval of iterations
        of the map to be plotted.
    start_point: list of length 2 [x, y] specifying the x and y coordinates of the point from which to
        begin iteration
    orbit_color: the color of the orbit plot, default is red.
    """
    
    curr_point = start_point
    visited_points = [start_point]
    n = len(nrule)
    m = len(X_m) # this is m-1 because we include x axis as a line
    
    # ensure basic setup is correct    
    assert m >= 2
    
    for i in nrule:
        assert 0 < i[0] <= 90
        assert i[1] == 0 or i[1] == 1
        assert type(i[2]) == int and 1 <= i[2] <= m
    
    # begin iteration
    for i in range(0, num_iter):
        # get rule information
        rule_num = i % n
        curr_rule = nrule[rule_num]
        step_ang = curr_rule[0]
        o = curr_rule[1]
        d = curr_rule[2]
        
        ################
        # From current point, order lines via distance, choose line at specified dist
        line_indx = 0
        line_dist_ordering = []
        for j in X_m:
            d_j = dist_point_line(j, curr_point)
            line_dist_ordering.append(d_j)
        # add distance from x-axis
        line_dist_ordering.append(abs(curr_point[1]))
        
        for j in range(0, d):
            # iterate through, set mins to large value to eventually reach
            # specified distance value
            min_indx = line_dist_ordering.index(min(line_dist_ordering))
            line_dist_ordering[min_indx] = 1000000000

        # determine index of line to map to    
        min_indx = line_dist_ordering.index(min(line_dist_ordering))
        line_indx = min_indx
        
        if(line_indx == m):
            # then is mapping to x-axis
            mapping_line1 = [0, 0]
            mapping_line2 = [0, 0]
        else:  
            mapping_line1 = X_m[line_indx]
            s = math.tan(math.radians(mapping_line1[0]))
            mapping_line2 = (s, mapping_line1[1])
        ################      
        
        # upon picking line, determine two possible step lines via mapping angles,
        # and pick one given the orientation.
        possible_lines = mapping_options(mapping_line1[0], step_ang)
        iteration_line_slope = possible_lines[o]
        
        # generate the line to find the next point in the iteration of the map
        iteration_line = [iteration_line_slope, newB(iteration_line_slope, curr_point)]

        # determine next point in the orbit, add to visited points
        next_point = inter(iteration_line, mapping_line2)
        curr_point = next_point
        visited_points.append(curr_point)
    
    # now plot the lines in X_m, and the orbit generated by the n-rule map in X_m
    # first plot the m lines
    xLine = np.linspace(-400, 400, 1000)
    plt.hlines(0, -400, 400, linewidth=2)
    
    for k in X_m:
        s = k[0]
        s = math.tan(math.radians(s))
        b = k[1]
        plt.plot(xLine, (s*xLine + b), color='black', linewidth=2)
    
    # plot the orbit    
    x = []
    for i in visited_points[points_to_plot[0]:points_to_plot[1]]:
        x.append(i[0])
    y = []
    for i in visited_points[points_to_plot[0]:points_to_plot[1]]:
        y.append(i[1])
    
    plt.plot(x, y, color=orbit_color)
    plt.xlim(-8, 16)
    plt.ylim(-8, 16)
    plt.gca().set_aspect('equal', adjustable='box')
    
    plt.draw
    
    # return period of orbit, if found:
    xt = []
    for i in visited_points[100:]:
        xt.append(i[0])
    
    yt = []
    for i in visited_points[100:]:
        yt.append(i[1])
        
    periodic_pointsx = list_duplicates_of(xt, xt[-1])
    periodx = periodic_pointsx[-1] - periodic_pointsx[-2]
    
    periodic_pointsy = list_duplicates_of(yt, yt[-1])
    periody = periodic_pointsy[-1] - periodic_pointsy[-2]
    if(periodx == periody):
        print("The periodic orbit has period of ", periodx)
    else:
        print("The periodic orbit has period of ", max([periodx, periody]))
        
    return visited_points

File: NRuleMaps.jl
### A Pluto.jl notebook ###
# v0.12.17

using Markdown
using InteractiveUtils

# ╔═╡ b26a3f0c-448e-11eb-1cf0-25438afc670d
using Plots

# ╔═╡ 98aeef06-44a5-11eb-1095-adc7773fd362
using LinearAlgebra

# ╔═╡ 9e6d6934-448e-11eb-21f3-69467bda2410
function lineAngle(slope1, slope2)
    a = slope2 - slope1
    b = 1.0 + (slope2*slope1)
    c = abs(a / b)
    d = atan(c)
    return d
end

# ╔═╡ c3fb3d48-448e-11eb-3dff-6fccd893312c
function lineZero(m, b)
    # find root of the passed line
    x = (-b)/m
    y = 0
    return [x, y]
end

# ╔═╡ c3fbc362-448e-11eb-36b3-a5c2694955da
function findFunc(point1, point2)
    # find the linear function that fits two points, returns m, b
    a = point2[2] - point1[2]
    b = point2[1] - point1[1]
    c = (point2[1] * point1[2]) - (point1[1] * point2[2])
    m = a/b
    d = c/b
    return [m, d]
end

# ╔═╡ c3fc5fd4-448e-11eb-2347-9ddee5b2cc8a
function pFunc(m)
    # find the function perpindicular to passed function flope
    if m != 0.0
        mperp = -1.0/m
    end
    if m == 0
        mperp = 0.0
    end

    return mperp
end

# ╔═╡ c40b9c4c-448e-11eb-1cbc-45324d95420d
function newB(m, point)
    # find the y-intersection that the function point pair pass through
    a = m*point[1]
    b = point[2] - a
    return b
end

# ╔═╡ c41174be-448e-11eb-334d-f97165d225d0
"""Find the distance between points p1 = [x, y] and p2"""
function dist(p1, p2)
    # find the distance between two points
    p = p1-p2
    d = norm(p)
    return d
end

# ╔═╡ c418c974-448e-11eb-01b9-8167c73c6e24
"""Find intersection point between lines of form [slope, intercept]"""
function inter(mb1, mb2)
    # find line intersection point between two lines
    a = mb2[2] - mb1[2]
    b = mb1[1] - mb2[1]
    
    if b==0
        return 0, mb1[2]
    end
        
    x = a / b
    y = (mb1[1] * x) + mb1[2]
    return [x, y]
end

# ╔═╡ c41eb1ec-448e-11eb-288d-eb4fecc22f89
"""Find the distance between a line [slope, intercept] and a point [x, y], 
where the slope of the line is an angle in degrees."""
function distpointline(line, point)
    s = tan(deg2rad(line[1]))
    p = pFunc(s)
    perp_line = [p, newB(p, point)]
    # now find intersection of the two lines, where lp is not an angle but usual slope
    lp = (s, line[2])
    # find line intersection point
    intersection = inter(perp_line, lp)
    # determine distance and return
    d = dist(point, intersection)
    return d
end

# ╔═╡ c425aa88-448e-11eb-067f-132bc8887e8a
"""Takes the target lineslope angle, and the given rule step angle,
    and returns the two lines that intersect at the target line at the step angle
"""
function mappingOptions(target_line_angle, step_angle)
    # construct base step angle slopes
    b = tan(deg2rad(step_angle + target_line_angle))
    b2 = tan(deg2rad(target_line_angle - step_angle))
    lines = [b, b2]
    return lines
end

# ╔═╡ 2b64bd10-448f-11eb-361c-1798fccb5fde
# plot the visited points
function plotOrbit(X_m, visitedPoints, points_to_plot, orbit_color)
    x = -800:800
    plot(x, zeros(1601), color="black", xlims=(-20,20), ylims=(-20, 20))

    # now plot lines composing X_m
    @simd for k in X_m
        # define y coordinates
        m = k[1]
        m = tan(deg2rad(m))
        b = k[2]
        plot!(x, (x*m).+b, color="black")
    end

    # now plot the orbit
    tx = []
    for i in visitedPoints[points_to_plot[1]+1:points_to_plot[2]+1]
        push!(tx, i[1])
    end
    ty = []
    for i in visitedPoints[points_to_plot[1]+1:points_to_plot[2]+1]
        push!(ty, i[2])
    end
    # plot
    plot!(tx, ty, color=orbit_color, label="Orbit")
end

# ╔═╡ ef6fb954-448e-11eb-0915-299e31bffd95
"""
The following code takes an n-rule, and plots the path it generates
    in the specified space X_m.  If the n-rule map converges to a periodic orbit
    within num_iter iterations, the program will return the period.
    Inputs:
    X_m: input a list of m tuples, where m is the number of lines plus the x axis, so 
        X_m is of form [(slope1, intercept1), (slope2, intercept2), ...] for line slope and intercept
        where line slope is an angle in degrees.  Further, we require that len(X_m) >= 2
    n-rule: takes a list of n length three tuples of form 
        [(angle1, orientation1, distance1), (angle2, orientation2, distance2), ...]
        where each angle parameter is between 0 and 90, orientation is either 1 or 0, and distance 
        is a positive integer from 1 to m inclusive (dealing with m+1 lines as add the x axis).  If angle=90, orientation can be either 0 or 1 with no change
        of map behavior
    num_iter: number of times to iterate the map
    points_to_plot: an interval [start_point, end_point] that determines the interval of iterations
        of the map to be plotted.
    start_point: list of length 2 [x, y] specifying the x and y coordinates of the point from which to
        begin iteration
    orbit_color: the color of the orbit plot, default is red.
"""
function nRuleMap(X_m, nrule, num_iter, points_to_plot, start_point, orbit_color="red")
    currPoint = start_point
    visitedPoints = [start_point]
    n = length(nrule)
    m = length(X_m) # m-1 as we include x-axis as a line
    
    # begin iteration
    for i in 1:num_iter
        # get rule information
        rulenum = (i%n) + 1
        currRule = nrule[rulenum]
        step_ang = currRule[1]
        o = currRule[2]
        d = currRule[3]
        
        # order liens via distance and choose line at specified distance
        lineIndx = 1
        lineDistOrdering = []
        for j in X_m
            d_j = distpointline(j, currPoint)
            push!(lineDistOrdering, d_j)
        end
        
        # add distance from x axis
        xdist = abs(currPoint[2])
        push!(lineDistOrdering, xdist)
        
        @simd for j in 1:d-1
            # we must find the index of the line in X_m
            # with the dth closest line, so set all other lines
            # to the max distance
            minIndx = argmin(lineDistOrdering)
            lineDistOrdering[minIndx] = Inf
        end
        
        # now determine the index of the line to map to
        lineIndx = argmin(lineDistOrdering)
        # now if index is m+1, then it is the x axis, as we added that to the end
        if lineIndx == m+1
            mappingLine1 = [0, 0]
            mappingLine2 = [0, 0]
		else
			mappingLine1 = X_m[lineIndx]
			s = tan(deg2rad(mappingLine1[1]))
			mappingLine2 = (s, mappingLine1[2])
		end
		
		# upon picking a line, determine two possible step lines via mapping angles
		# and pick one given the orientation
		possibleLines = mappingOptions(mappingLine1[1], step_ang)
		iterationLineSlope = possibleLines[o+1]
		
		# generate the line to find the next point in iteration
		iterationLine = [iterationLineSlope, newB(iterationLineSlope, currPoint)]
		
		# determine the next point in the orbit and add to visited points
		nextPoint = inter(iterationLine, mappingLine2)
		currPoint = nextPoint
		push!(visitedPoints, currPoint)
	end
	
	## Now we plot the lines composing X_m and the iterations of the map
	# first plot the x axis. We plot lines on fairly wide axis
    plotOrbit(X_m, visitedPoints, points_to_plot, orbit_color)
end

# ╔═╡ 0e77beee-449b-11eb-39dd-6f2a6335dee0
"""
A symbolic implementation of n-rule maps. The function symbolicnrulemap takes an n-rule, and plots the path it generates in the specified space X_m.  If the n-rule map converges to a periodic orbit within num_iter iterations, the program will return the period.
Inputs:
X_m: input a list of m tuples, where m is the number of lines plus the x axis, so 
	X_m is of form [(slope1, intercept1), (slope2, intercept2), ...] for line slope 	and intercept where line slope is an angle in degrees.  Further, we require that 	len(X_m) >= 2.  The function is designed so that the index in the list X_m is the symbol value of the corresponding line, and m+1 index is the symbol value of the x-axis.
n-rule: takes a list of n length three tuples of form 
	[(angle1, orientation1, distance1), (angle2, orientation2, distance2), ...]
	where each angle parameter is between 0 and 90, orientation is either 1 or 0, and distance is a positive integer from 1 to m+1 inclusive.  If angle=90, orientation can be either 0 or 1 with no change of map behavior
num_iter: number of times to iterate the map
points_to_plot: an interval [start_point, end_point] that determines the interval of iterations
	of the map to be plotted.
start_point: list of length 2 [x, y] specifying the x and y coordinates of the point from which to
	begin iteration
orbit_color: the color of the orbit plot, default is red.
"""
function symbolicnRuleMap(X_m, nrule, num_iter, points_to_plot, start_point, orbit_color="red")
	
	currPoint = start_point
    visitedPoints = [start_point]
    n = length(nrule)
    m = length(X_m) # m-1 as we include x-axis as a line
    
    # begin iteration
    for i in 1:num_iter
        # get rule information
        rulenum = (i%n) + 1
        currRule = nrule[rulenum]
        step_ang = currRule[1]
        o = currRule[2]
        d = currRule[3]
        
        # determine line
        if d == m+1
            mappingLine1 = [0, 0]
            mappingLine2 = [0, 0]
		else
			mappingLine1 = X_m[d]
			s = tan(deg2rad(mappingLine1[1]))
			mappingLine2 = (s, mappingLine1[2])
		end
		
		# upon picking a line, determine two possible step lines via mapping angles
		# and pick one given the orientation
		possibleLines = mappingOptions(mappingLine1[1], step_ang)
		iterationLineSlope = possibleLines[o+1]
		
		# generate the line to find the next point in iteration
		iterationLine = [iterationLineSlope, newB(iterationLineSlope, currPoint)]
		
		# determine the next point in the orbit and add to visited points
		nextPoint = inter(iterationLine, mappingLine2)
		currPoint = nextPoint
		push!(visitedPoints, currPoint)
	end
	
	## Now we plot the lines composing X_m and the iterations of the map
	# first plot the x axis. We plot lines on fairly wide axis
    plotOrbit(X_m, visitedPoints, points_to_plot, orbit_color)
end

# ╔═╡ Cell order:
# ╠═b26a3f0c-448e-11eb-1cf0-25438afc670d
# ╠═98aeef06-44a5-11eb-1095-adc7773fd362
# ╠═9e6d6934-448e-11eb-21f3-69467bda2410
# ╠═c3fb3d48-448e-11eb-3dff-6fccd893312c
# ╠═c3fbc362-448e-11eb-36b3-a5c2694955da
# ╠═c3fc5fd4-448e-11eb-2347-9ddee5b2cc8a
# ╠═c40b9c4c-448e-11eb-1cbc-45324d95420d
# ╠═c41174be-448e-11eb-334d-f97165d225d0
# ╠═c418c974-448e-11eb-01b9-8167c73c6e24
# ╠═c41eb1ec-448e-11eb-288d-eb4fecc22f89
# ╠═c425aa88-448e-11eb-067f-132bc8887e8a
# ╠═2b64bd10-448f-11eb-361c-1798fccb5fde
# ╠═ef6fb954-448e-11eb-0915-299e31bffd95
# ╠═0e77beee-449b-11eb-39dd-6f2a6335dee0
